torch/fx/experimental/Partitioner.py

```diff
@@ -0,0 +1,151 @@
+        self.parents.append(partition)
+
+    def add_child(self, partition: 'Partition') -> None:
+        self.children.append(partition)
+
+    def get_input_nodes(self) -> [int]:
+        """input nodes are coming from two places:
+        placeholder or output from its parents output
+        return a list of node indexes
+        """
+        input_nodes: List[Tuple[int, str, str]] = []
+        for node_index, node_name, node_op in self.node_list:
+            if node_op == 'placeholder':
+                input_nodes.append((node_index, node_name, node_op))
+        for partition_id in self.parents:
+            input_nodes.append(partition.get_output_node())
+        return input_nodes
+
+    def get_output_node(self) -> int:
+        """the output node of this partition should be the last node of self.node_list
+        """
+        assert len(self.node_list) != 0
+        return self.node_list[-1]
+
+    def __str__(self) -> str:
+        line: str = 'partition id: ' + str(self.partition_id) + '\n'
+        line += 'all nodes in this partition: '
+        for node_index, node_name, node_op in self.node_list:
+            line += '(node_' + str(node_index) + ':' + node_name +  ',' + node_op + ') '
+        line += '\n'
+        line += 'parent partitions:'
+        for parent in self.parents:
+            line += 'partition_' + str(parent.partitition_id) + ' '
+        line += '\n'
+        line += 'children partitions:'
+        for child in self.children:
+            line +=  str(child.partition_id) + ' '
+        line += '\n'
+        line += 'input nodes: '
+        input_nodes = self.get_input_nodes()
+        for node_index, node_name, node_op in input_nodes:
+            line += '(node_' + str(node_index) + ':' + node_name +  ',' + node_op + ') '
+        line += '\n'
+        output_node = self.get_output_node()
+        line += 'output node: ' + '(node_' + str(output_node[0]) + ':' + output_node[1] + ',' + output_node[2] + ')\n'
+        return line
+
+class Partitioner:
+    def __init__(self) -> None:
+        self.partitions: List[Partition] = []
+
+    """
+    After a partitioner is created,
+    it first check if multiple types of devices (backends) are involved
+    So far, we assume that there is only one backend and one device with unlimited memory
+    """
+
+    def partition_graph(self, fx_module: GraphModule, devices: List[str]) -> None:
+        graph: Graph = fx_module.graph
+        """check if devices are different backend types"""
+        multiple_backend_types: bool = self.check_multiple_backend_types(devices)
+        """one node could be partitioned into multiple partitions
+           mapping {node_index: [partition_ids]}
+        """
+        node_to_partition_mapping: Dict[int, List[int]] = {}
+        """only one type of devices"""
+        if not multiple_backend_types:
+            node_to_partition_mapping = self.load_balanced_partition(graph, devices)
```

Comment:
LoadBalanced and SinglePartition are both valid find_partition methods, They should both be routed to from partition_graph.